Waypoints navigation fleshed out, but not working. PR for review

markmellors opened the pull request:
I'm happy with the waypoint navigation code structure. I've integrated it into main to test it (not in an example challenge yet).
It doesn't work in weird ways though. sometimes the robot does drive autonomously when in the `AUTO` state, sometimes not. the screen displays the status, though the status doesn't always change as expected (e.g. waypoint heading doesn't update as the robot is rotated).  The robot status pose orientation doesn't appear to update either.

I suspect I've done multiple things wrong...

markmellors committed:
PID class created

markmellors committed:
navigation file created, with headingToWaypoint function

markmellors committed:
PID moved to its own file

markmellors committed:
WaypointNav class created

markmellors committed:
added routes definitions

markmellors committed:
renamed navigation

markmellors committed:
trial, minimal implementation of waypoint navigation. seems to work?

markmellors committed:
updated screen and status for waypoints

markmellors committed:
some debugging, still not working

markmellors commented:
update: `_status.pose.heading` appears to not update when in running mode, so maybe something mroe fundamental isn't working on master

markmellors commented:
another observation: the IMU data (orientation) looks correct in the log. but the odometry pose (heading) doesn't, its a constant 0.0.
edit: that's because `Status::updatePose` didn't update the heading like it used to,
see: https://github.com/robberwick/tauradigm-teensy/blob/afa439d565e2557bf05fed206355be0946d35cdf/src/main.cpp#L119
adding `    pose.heading = heading;`  fixed all the heading issues. 
waypoint navigation appears to be working now (on the bench), but a code review would still be useful

markmellors committed:
fixed odometry pose updating error

robberwick commented:
No worries.. I'll check out the code and have a look, then if I identify any changes that might want to be made, I can submit a PR against your branch 👍

markmellors commented:
Will and I spotted a bug today in my PID routine - it needs a special case adding for wrapping targets, such as a heading. E.g. target heading = 0, actual heading =6 should give a small (<1) error, not 6 (when heading is in radians)

markmellors committed:
pi wrapping ability added to PID and used for waypoint heading PID

markmellors commented:
another bug spotted and squashed. The order of the elements of `Pose` had changed, which meant my routes were now defining `heading` and `x` , not `x` and `y`. ...

markmellors committed:
route definitions updated to reflect pose element order, more variables displayed on screen for debugging

markmellors commented:
I've realised my attempt at maintaining both a  'waypoint coordinate system' and an 'abosolute coordinate system' (robotStatus.pose)  was flawed - you can't jsut apply an x,y and heading offset, any change in x or Y also needs to have the heading offset taken into consideration. so for now, I've taken that out, and there's only one coordinate system.

markmellors commented:
I think I spotted my main bug: I was mixing up a local variable with a class property, this one: https://github.com/robberwick/tauradigm-teensy/blob/b61f01b8e551ce2f5633429eb0b330627e640927/src/waypoint_navigation.cpp#L27
so the target location as far as everything else was concerned was always 0,0. which explains why the robot kept driving in circles around the start location!

markmellors commented:
another bug spotted and fixed, this is one thats been in the code for ages, and has been causing weird, sporadic issues with everything:
https://github.com/robberwick/tauradigm-teensy/blob/54e6d28d8b618e5c4b3b39b9636a0b503b71d1c6/src/robot_hal.cpp#L88
was intended to be true if either command was non zero. either being another way of saying OR, not AND.......

that explains why the robot sometimes totally stopped when the target heading was at 90degrees, as that caused the commanded speed for one wheel to be zero, then this bug mistakenly resulted in both motors being set to zero.

markmellors commented:
another bug found from master. The odometry is only updated everytime the sensor data is sent, but the encoders are updated every encoder read. the odometry uses the difference between latest and last-but-one encoder reads to work out distance travelled. If there's been numerous enocder reads inbetween, ticks are lost

markmellors committed:
removed waypoint coordinate system, fixed:
feedforward one motor stopped bug
local vs class level targetwaypoint variable bug
encoder counts getting lost bug
waypoint heading discontinuity
added heading offset

markmellors commented:
works now!
